src/subgraph/convolution-2d.c

```diff
@@ -561,6 +561,11 @@ static inline enum xnn_compute_type validate_datatypes_with_bias(
           output_datatype == xnn_datatype_fp32)
       {
         return xnn_compute_type_fp32;
+      } else if (input_datatype == xnn_datatype_fp16 &&
+          bias_datatype == xnn_datatype_fp32 &&
```

Comment:
Are you certain that the bias in this case is fp32? Typically for fp16 inference all float32 are transformed to float16. The microkernels only support fp16 bias: https://github.com/google/XNNPACK/blob/6ef9832d025fb385d69c8fec88e0cb6ffa950e7f/src/f16-igemm/gen/f16-igemm-1x16-minmax-neonfp16arith-ld64.c#L48

The bias is loaded here.

Reply:
Ignore this, the bias and weights are converted in the packing function.

This is inefficient, we should optionally be able to pass fp16 weights and biases. Do you fancy adding this in a separate PR? I think only some small changes to this file are necessary, at the operator level fp16 weights and biases are handled.

Reply:
Yeah, I noticed that as well. I'll create a follow-up PR to allow passing fp16 weights for conv & fully-connected.